File: code/remove_nth_node_from_end.py
# https: // leetcode.com/problems/remove-nth-node-from-end-of-list/

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next


class Solution:
	def removeNthFromEnd(self, head: ListNode, n: int) -> ListNode:
		tmp = curr = head
		size = 0
		while (tmp):
			tmp = tmp.next
			size += 1
		if (size - n <= 0):
			return head.next
		for _ in range(size - n - 1):
			curr = curr.next
		curr.next = curr.next.next
		return head

	def removeNthFromEnd2(self, head: ListNode, n: int) -> ListNode:
		fast = slow = head
		for _ in range(n):
			fast = fast.next
		if (fast is None):
			return head.next
		while fast.next:
			fast = fast.next
			slow = slow.next
		slow.next = slow.next.next
		return head

File: code/remove_nth_node_from_end.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

// My first c solution. Couldn't come with the solution without using "call by reference"

struct ListNode* removeNthFromEnd(struct ListNode* head, int n){
    struct ListNode* curr = head;
    int size = 0;
    int counter;

    while (curr)
    {
        curr = curr->next;
        size++;
    }
    if (size - n <= 0)
    {
        head = head->next;
        return (head);
    }
    struct ListNode *tmp = head;
    counter = size - n - 1;
    while (counter)
    {
        tmp = tmp->next;
        counter--;
    }
    tmp->next = tmp->next->next;
    return head;
}
